library/dd-sdk-android-session-replay/src/test/kotlin/com/datadog/android/sessionreplay/recorder/ScreenRecorderTest.kt

```diff
@@ -51,234 +61,372 @@ internal class ScreenRecorderTest {
-            window = mockWindow,
-            defaultWindowCallback = null
-        )
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // When
-        testedRecorder.startRecording(mockActivity)
+        testedRecorder.stopRecording()
 
         // Then
-        val captor = argumentCaptor<Window.Callback>()
-        verify(mockWindow).callback = captor.capture()
-        assertThat(captor.firstValue).isInstanceOf(RecorderWindowCallback::class.java)
-        assertThat((captor.firstValue as RecorderWindowCallback).wrappedCallback)
-            .isInstanceOf(NoOpWindowCallback::class.java)
+        fakeWindowsList.forEach {
+            val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
+            verify(it.decorView.viewTreeObserver).addOnDrawListener(captor.capture())
+            verify(it.decorView.viewTreeObserver).removeOnDrawListener(captor.firstValue)
+        }
+        assertThat(testedRecorder.windowsListeners).isEmpty()
     }
 
     @Test
-    fun `M do nothing W startRecording() { activity window is null }`() {
+    fun `M unregister first and clean the listeners W startRecording()`() {
         // Given
-        val mockActivity: Activity = mock()
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // When
-        testedRecorder.startRecording(mockActivity)
+        testedRecorder.startRecording(fakeWindowsList, mockActivity)
 
         // Then
-        assertThat(testedRecorder.drawListeners).isEmpty()
+        fakeWindowsList.forEach {
+            val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
+            it.decorView.viewTreeObserver.inOrder {
+                verify().addOnDrawListener(captor.capture())
+                verify().removeOnDrawListener(captor.firstValue)
+                verify().addOnDrawListener(captor.capture())
+            }
+        }
     }
 
     @Test
-    fun `M unregister the RecorderOnDrawListener W stopRecording()`(forge: Forge) {
+    fun `M register listener startRecording() { more activities }`(forge: Forge) {
         // Given
-        val mockViewTreeObserver: ViewTreeObserver = mock()
-        val mockActivity = mockActivity(forge, mockViewTreeObserver)
-        testedRecorder.startRecording(mockActivity)
+        val fakeWindowsActivityPairs = forge.aList {
+            aMockedWindowsList() to aMockedActivity()
+        }
 
         // When
-        testedRecorder.stopRecording(mockActivity)
+        fakeWindowsActivityPairs.forEach {
+            testedRecorder.startRecording(it.first, it.second)
+        }
 
         // Then
-        val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
-        verify(mockViewTreeObserver).addOnDrawListener(captor.capture())
-        verify(mockViewTreeObserver).removeOnDrawListener(captor.firstValue)
+        fakeWindowsActivityPairs.map { it.first }.flatten().forEach {
+            val captor = argumentCaptor<ViewTreeObserver.OnDrawListener>()
```

Comment:
We don't need captor here, because we don't run any assertions on the values captured. We can use simply `verify().addOnDrawListener(any())`

Reply:
well actually I am doing it here:
`verify(it.decorView.viewTreeObserver).removeOnDrawListener(captor.firstValue)`. I know it could work with `any` too but I would rather use an explicit verification.

Reply:
Nevermind, you are right, I was looking at the wrong line actually.